### ISSUE TITLE
Loss calculation and updating weights

### ISSUE THREAD

**Issue by ZheMann** [rkfg/gpt-2#7]
@rkfg kinda embarrassing to ask this after working with GPT-2 for a couple of weeks already, but here it goes. I thought I had it al clear in my mind, but as I started thinking about it: how does the loss actually gets calculated while finetuning in an unsupervised way? And how do the weights get updated?  The longer I think about it, the more I start doubting about whether I fully understand the training code. 

If it is supervised, I can imagine it works as follows:
While training, it take a sample of length defined in `hparams.json`, and predicts the next word. Then, the cross-entropy is calculated based on the predicted next word and the actual next word that should occur. 

Is this correct? And based on the cross-entropy SGD/Adam is performed?

Sorry to bother you, but hopefully you have some spare time to give a short explanation.

**Comment by rkfg** [comment]
I couldn't fully get my head around this but two things are certain:
1. The training code always sample 1024 tokens, no matter how long the context is. It's hardcoded in `sample_batch()` in `train.py`.
1. The loss is calculated for the last token `n` given the previous `n-1` tokens. That's what the comment in `src/model.py` states almost at the end: `# Language model loss.  Do tokens <n predict token n?` (for 1024 tokens that means we predict 1024th token from the first 1023). I don't understand the logic of these matrix multiplications and reshaping so can only guess from the few comments they left. The predicted distribution across your `N` dictionary tokens is then compared with the one-hot vector of the same size where 1 is set for the 1024th token and all other positions are zero. The difference between these vectors is your loss (the cross-entropy formula is actually very simple and easy to understand) and it's used to update the weights.

**Comment by ZheMann** [comment]
Massive thank you for the quick reply! I was hoping for an answer like this, as this makes sense. 

Last question: we have a dictionary consisting of `N` tokens. A token in the dictionary can be a character, subwords or a complete word, right? So, for the sake of illustration: assume the word 'uninterested' does not occur in our dictionary but has to be constructed from subwords 'un', 'interest' and 'ed'. Is this trained in the way you described? So during the first training step, based on 1023 tokens, the subword 'un' is predicted, The next training step, based on the previous 1023 tokens (including subword 'un'), the subword 'interest' is predicted, and so on?

The way the model learns makes sense to me now, I only have some thoughts about how this work with subwords.

**Comment by rkfg** [comment]
>So during the first training step, based on 1023 tokens, the subword 'un' is predicted

Correct.

>The next training step, based on the previous 1023 tokens (including subword 'un'), the subword 'interest' is predicted, and so on?

Incorrect. The sampling is always random so that each part of the corpus can be used without excessive overfitting. Let's speculate about what happens if your idea is implemented.

Say, you have a corpus of books concatenated altogether. If you go through it sequentially the model would bias towards this first book (albeit slightly but still). Then you get to the second book and its structure and words are different from the first. You'll get a bit of a loss spike. It's not fatal but completely unnecessary as you'll be losing time because of these spikes. It's much better if the model gets random pieces of all sorts (be it fiction, documental books, articles, descriptions, dialogs and so on) and learns to predict different things each time. I suppose the process will be faster just as the original authors say in the paper that training the model on diverse texts yields better results than training just on books or Wikipedia articles. You can easily change the sampling logic in `src/load_dataset.py`, `sample(self, length)` function. Instead of random use a counter or something and observe the loss behavior.

**Comment by ZheMann** [comment]
Well this all makes sense. I did have a global idea about how it worked, but now also the details become more and more clear to me. 

One thing to make sure if I understand it correctly: in the context of taking a sample of 1024 tokens while training, tokens are single characters (or spaces) right?  Because sometimes the word token is also used to indicate words.

**Comment by rkfg** [comment]
A token has a representation of subword unit but internally it's just a number from 0 to `N`. You can see the tokens in `sp.vocab` where `▁` means space. Sometimes it's a whole word if it's common enough in your corpus to get itself a dedicated token.

**Comment by ZheMann** [comment]
Perfect, thank you so much! You really helped me out, I really appreciate!

**Comment by rkfg** [comment]
Just to add, those numbers in `sp.vocab` really represent the tokens used in encoding but they're shifted by 5 or 6 (don't remember exactly but probably 5 as it's the number of internal tokens added by the library like `<unk>`, `<s>` and `</s>` + my own custom tokens for end of line and end of text) and for some reason negative. So your first real token would be `5` (`-0` in the dictionary) and the last one is `N` + 5. Encode a piece of text using `spm_encode` with your dictionary and look the numbers up in this file manually (deducing 5), should work. You're welcome!
